python/ray/_private/runtime_env/validation.py

```diff
@@ -65,184 +214,108 @@ class RuntimeEnvDict:
+    task, and the current runtime env comes from the current TaskSpec.
+
+    The child's runtime env dict is merged with the parents via a simple
+    dict update, except for runtime_env["env_vars"], which is merged
+    with runtime_env["env_vars"] of the parent rather than overwriting it.
+    This is so that env vars set in the parent propagate to child actors
+    and tasks even if a new env var is set in the child.
+
+    Args:
+        runtime_env_dict (dict): A runtime env for a child actor or task.
+
+    Returns:
+        The resulting merged JSON-serialized runtime env.
+    """
+    assert child_runtime_env is not None
+    assert parent_runtime_env is not None
+
+    result = copy.deepcopy(child_runtime_env)
 
     # If per-actor URIs aren't specified, override them with those in the
     # job config.
-    if "uris" not in runtime_env_dict and "uris" in parent_runtime_env:
-        runtime_env_dict["uris"] = parent_runtime_env.get("uris")
-
-    return runtime_env_dict
-
-
-def parse_conda_str(conda: str):
-    yaml_file = Path(conda)
-    conda_dict = dict()
-    if yaml_file.suffix in (".yaml", ".yml"):
-        if not yaml_file.is_file():
-            raise ValueError(f"Can't find conda YAML file {yaml_file}")
-        try:
-            conda_dict = yaml.safe_load(yaml_file.read_text())
-        except Exception as e:
-            raise ValueError(f"Invalid conda file {yaml_file} with error {e}")
-        return conda_dict
-    else:
-        logger.info(f"Using preinstalled conda environment: {conda}")
-        return conda
-
-
-def parse_pip_str(pip: str):
-    # We have been given a path to a requirements.txt file.
-    pip_file = Path(pip)
-    if not pip_file.is_file():
-        raise ValueError(f"{pip_file} is not a valid file")
-    return pip_file.read_text().splitlines()
-
-
-def parse_pip_and_conda(runtime_env):
-    if runtime_env is not None:
-        new_runtime_env = runtime_env.copy()
-        if isinstance(new_runtime_env.get("pip"), str):
-            new_runtime_env["pip"] = parse_pip_str(new_runtime_env["pip"])
-        if isinstance(new_runtime_env.get("conda"), str):
-            new_runtime_env["conda"] = parse_conda_str(
-                new_runtime_env["conda"])
-    else:
-        new_runtime_env = None
-    return new_runtime_env
+    if "uris" not in result and "uris" in parent_runtime_env:
+        result["uris"] = parent_runtime_env.get("uris")
+
+    # Override environment variables.
+    result_env_vars = copy.deepcopy(parent_runtime_env.get("env_vars") or {})
+    child_env_vars = child_runtime_env.get("env_vars") or {}
+    result_env_vars.update(child_env_vars)
+    if len(result_env_vars) > 0:
+        result["env_vars"] = result_env_vars
+
+    # NOTE(architkulkarni): This allows worker caching code in C++ to
+    # check if a runtime env is empty without deserializing it.
+    assert all(val is not None for val in result.values())
+
+    return result
```

Comment:
I think right now the behavior in the implementation doesn't quite match what's in the docstring, because the dicts don't seem to be merged here.  For example, if the parent has a "pip" field but the child only specifies an "env_vars" field, the "pip" field won't make appear in `result`, but I think we want it to.

Reply:
Yep you're right. We clearly didn't have unit tests for this :(

Reply:
Yeah, I thought it was at least covered in some of the end-to-end tests in `test_runtime_env_complicated.py` but it's possible it wasn't :(

Reply:
added the logic & tests